changelog/329.feature.md

```diff
@@ -0,0 +1,36 @@
+Extend the `FormValidation` action with support for extracting slots. If you
+want to extract a custom slot, add the slots name to the list of `required_slots`
+and add a method `extract<slot name>` to your action:
+
+```python
+from typing import Text, Dict, Any, List, Optional
+
+from rasa_sdk.forms import (
+    FormValidationAction,
+)
+
+class FormWithSlotExtractions(FormValidationAction):
+    def name(self) -> Text:
+        return "some_form"
+
+    async def required_slots(
+        self,
+        dispatcher: "CollectingDispatcher",
+        tracker: "Tracker",
+        domain: "DomainDict",
+    ) -> Optional[List[Text]]:
+        return ["my slot"]
```

Comment:
in which situations would you need to fill this out? you can specify slots without a slot mapping in the domain right?

Reply:
ok i just realised why, in case you want to dynamically change the list of requested slots :D

Reply:
you'd override for two use cases:
* dynamic slots
* custom slot mappings

Reply:
ok so the custom slot mapping part i'm confused about: if you specify your slots without a slot mapping in the domain file, shouldn't it just use the `extract_` method for that slot?

Reply:
You mean that we would just try to extract every slot for which we have an `extract_` function defined? I think that would work fine if the slot mappings are static. It would become tricky in case your `required_slots` change dynamically

Reply:
but would you list the "optional" required slots in the domain in the first place? if you didn't, i guess you wouldn't have the issue with the dynamically changing ones right?

Reply:
> but would you list the "optional" required slots in the domain in the first place? 

Nope, I wouldn't.

>  i guess you wouldn't have the issue with the dynamically changing ones right?

@samsucik interacted with users who had dynamically changing custom slot mappings. So that problem would still persist.

Reply:
ok i'm super confused now - the way to extract the slot changes? how would that even work?

Reply:
I don't think "my" users wanted to dynamically change how slots are extracted. They wanted to dynamically change the list of required slots, but extraction was either given right in the domain or implemented using `extract_{slot}` (in either case, the implementation didn't change). The way I understand it is that this new implementation should use `extract_{slot}` whenever it exists for the given slot, and use the "default" extraction approach otherwise.

Reply:
Sounds good for you @akelad ?

Reply:
yes, that makes sense for me! So just to summarise, for the requested slot method that means you only need to fill it out if you want to dynamically change requested slots, correct?

Reply:
Exactly. This includes the case that you want to ignore / override slot mappings defined in the domain (ideally they just leave the slot mappings empty in the domain)